server/base/src/main/java/org/apache/accumulo/server/conf/ZooBasedConfiguration.java

```diff
@@ -0,0 +1,284 @@
+    var snapshot = snapshotRef.get();
+    if (snapshot == null) {
+      return updateSnapshot().getDataVersion();
+    }
+    return snapshot.getDataVersion();
+  }
+
+  /**
+   * The update count is the sum of the change count of this configuration and the change counts of
+   * the parents. The count is used to detect if any changes occurred in the configuration hierarchy
+   * and if the configuration needs to be recalculated to maintain consistency with values in the
+   * backend store.
+   * <p>
+   * The count is required to be an increasing value.
+   */
+  @Override
+  public long getUpdateCount() {
+    long count = 0;
+    long dataVersion = 0;
+    for (AccumuloConfiguration p = this; p != null; p = p.getParent()) {
+      if (p instanceof ZooBasedConfiguration) {
+        dataVersion = ((ZooBasedConfiguration) p).getDataVersion();
+      } else {
+        dataVersion = p.getUpdateCount();
+      }
+      count += dataVersion;
+    }
+
+    log.trace("update count result for: {} - data version: {} update: {}", propCacheId, dataVersion,
+        count);
+    return count;
+  }
+
+  @Override
+  public AccumuloConfiguration getParent() {
+    return parent;
+  }
+
+  public PropCacheId getCacheId() {
+    return propCacheId;
+  }
+
+  @Override
+  public @Nullable String get(final Property property) {
+    Map<String,String> props = getSnapshot();
+    String value = props.get(property.getKey());
+    if (value != null) {
+      return value;
+    }
+    AccumuloConfiguration parent = getParent();
+    if (parent != null) {
+      return parent.get(property);
+    }
+    return null;
+  }
+
+  @Override
+  public void getProperties(final Map<String,String> props, final Predicate<String> filter) {
+
+    parent.getProperties(props, filter);
+
+    Map<String,String> theseProps = getSnapshot();
+
+    log.trace("getProperties() for: {} filter: {}, have: {}, passed: {}", getCacheId(), filter,
+        theseProps, props);
+
+    for (Map.Entry<String,String> p : theseProps.entrySet()) {
+      if (filter.test(p.getKey()) && p.getValue() != null) {
+        log.trace("passed filter - add to map: {} = {}", p.getKey(), p.getValue());
+        props.put(p.getKey(), p.getValue());
+      }
+    }
+  }
+
+  @Override
+  public boolean isPropertySet(final Property property) {
+
+    Map<String,String> theseProps = getSnapshot();
+
+    if (theseProps.get(property.getKey()) != null) {
+      return true;
+    }
+
+    return getParent().isPropertySet(property);
+
+  }
+
+  public Map<String,String> getSnapshot() {
+    if (snapshotRef.get() == null) {
+      return updateSnapshot().getProps();
+    }
+    return snapshotRef.get().getProps();
+  }
+
+  @Override
+  public void invalidateCache() {
+    snapshotRef.set(null);
+  }
+
+  private final Lock updateLock = new ReentrantLock();
+
+  private @NonNull PropSnapshot updateSnapshot() throws PropStoreException {
+
+    PropSnapshot localSnapshot = snapshotRef.get();
```

Comment:
This class has a reference to `ZooStore`, which uses `Caffeine` to cache the entries from ZooKeeper. Then in addition to that, this class keeps another cached copy of the properties? Is that right?

Reply:
My initial thought was that this local copy may be unnecessary and add complexity with maintaining a two-tiered caching system. The Caffeine cache should be able to hold *all* of the properties, just need to make sure that the version is has is the latest. I'm wondering if this is a premature optimization and if the benefits outweigh the complexity cost.

Reply:
It's not really an attempt to prematurely optimize. Rather, it's an attempt to deal with existing issues. Some of them are:

1. We currently hold table configs forever. Even if we're a tserver no longer longer hosting any tablets from that table. Probably even if it is deleted. This can be a moderate memory leak if you routinely create many tables. To deal with this, the caffeine cache expires, and (hopefully) these configuration objects in the config hierarchy can be garbage collected (in the case of table/namespace config objects).
2. For existing code, we need to be able to detect when something changes in the configuration, like balancers, so we can reconfigure those classes, via the "Deriver" mechanism and tracking the `updateCount()` method. This implementation could change, but changing it too much might be out of scope for this PR.
3. If the caffeine cache expires something because of lack of activity, and we still have table/namespace/system configuration objects around, we need to be able to know when we need to reconfigure anything that might be using the Deriver mechanism.

A lot of this complexity isn't an attempt to optimize performance, so much as deal with these nuances. But yeah, there is some mindfulness to avoid some of the worst performing situations that introduces some complexity as well (constantly going back to ZK directly to check for changes, rather than relying on watchers, for example, is something we should avoid).

Reply:
The Caffeine cache has two timeout for cache eviction.  One is a hard eviction and the other is more "soft".

The hard eviction occurs if a cache item has not been accessed after the timeout period.  This allows the cache to evict items that are not being used - the use cases would be when a table migrates to another server - or if there are tables that are infrequently accessed.

The soft eviction is being leveraged to ensure eventual consistency without needing solely relying on watchers.  When the soft-timeout is reached and an item is accessed in the cache - the cache return the current value and then runs an async task to check / reload the value from the back end source of truth (ZooKeeper in this case).  Currently, that check compares the node version. If the version is the same, no other action is required. If the node changed, and the version does not match, the new value is loaded and an event is signaled so that the change is propagated.

The use case for this is for things like iterators - the iterator can keep accessing the cache value without pausing but be able to reach if a change not signaled by a watcher is detected.

Reply:
If the complexity of this two level scheme is there to deal with deleted tables another possible way to do deal with that is periodic reconciliation.  Some things in the Accumulo do this they just periodically run a task that looks for deleted things to clean out.  So could get the set of all tables id in ZK and then conceptually do something like `tableConfigCache.keySet().retainAll(tablesIdsFromZK)` every few hours AND maybe have a simpler once level cache.  However I am still looking at this, so not completely sure if my suggestion is any good.

> This can be a moderate memory leak if you routinely create many tables. To deal with this, the caffeine cache expires, and (hopefully) these configuration objects in the config hierarchy can be garbage collected (in the case of table/namespace config objects).

In the ServerConfigurationFactory class there is a static map of tableConfigs that may never get cleaned up in anyway.  So not sure if these could ever be GCed once created.

Reply:
I agree that there are opportunities for future clean-up.  This code tried to limit changes to existing behavior.  Once merged, additional PRs in future versions can be used to address more clean-up.  Those changes should be much smaller in scope and easier to deal with.